### PR TITLE
Inherit annotations s3

### LINKED ISSUE #375: allow inheriting annotations from base trait/classes
* add `@inherit` annotation working on def's
* extend collecting annotations in magnolia based on that -> `annotationsOf(Symbol)`

### PR DESCRIPTION
**Pull request by kubinio123**

### CODE REVIEW

File: src/core/macro.scala [hunk 1]
```diff
@@ -193,3 +144,99 @@ object Macro:
...
+        tpe.typeSymbol.annotations
+          .filter(filterAnnotation)
+          .map(_.asExpr.asInstanceOf[Expr[Any]])
+      }
+
+    def inheritedAnns: Expr[List[Any]] =
+      Expr.ofList {
+        tpe.baseClasses
+          .collect {
+            case s if s.name != tpe.typeSymbol.name => s.annotations
```
Comment (KacperFKorban) [inline src/core/macro.scala]: Comparing names is not 100% reliable, wouldn't comparing type symbols work?

File: src/core/macro.scala [hunk 2]
```diff
@@ -193,3 +144,99 @@ object Macro:
...
+          } // skip self
+          .flatten
+          .filter(filterAnnotation)
+          .map(_.asExpr.asInstanceOf[Expr[Any]])
+      }
+
+    def typeAnns: Expr[List[Any]] =
```
Comment (KacperFKorban) [inline src/core/macro.scala]: Could you add an end marker or curly braces to this def?

File: src/core/macro.scala [hunk 3]
```diff
@@ -193,3 +144,99 @@ object Macro:
...
+          (fromConstructor(tpe.typeSymbol) ++ fromDeclarations(tpe.typeSymbol))
+            .filter { case (_, anns) => anns.nonEmpty }
+        }
+      }
+
+    def inheritedParamAnns: Expr[List[(String, List[Any])]] =
+      Expr.ofList {
+        groupByParamName {
+          tpe.baseClasses
+            .filterNot(bc =>
```
Comment (KacperFKorban) [inline src/core/macro.scala]: Could you switch to curly braces here?

File: src/core/interface.scala [hunk 4]
```diff
@@ -1,9 +1,10 @@
 package magnolia1
 
-import language.experimental.macros
 import scala.annotation.tailrec
 import scala.reflect.*
 
+import language.experimental.macros
```
Comment (KacperFKorban) [inline src/core/interface.scala]: Does this import do anything anymore in Scala 3?

File: src/core/macro.scala (hunk 1 above)
Comment (kubinio123) [reply src/core/macro.scala]: Thanks, seems to work when I just compare `s` against `tpe.typeSymbol`

File: src/core/interface.scala (hunk 4 above)
Comment (kubinio123) [reply src/core/interface.scala]: I don't think so, not sure how it end up here :D

**Review by KacperFKorban — APPROVED** [review]
LGTM, but I would wait for @adamw approval as well :smile:

File: src/examples/show.scala [hunk 5]
```diff
@@ -66,13 +70,21 @@ trait GenericShow[Out] extends AutoDerivation[[X] =>> Show[Out, X]] {
...
         val annotationStr =
           if (anns.isEmpty) "" else anns.mkString("{", ",", "}")
 
         prefix(annotationStr, sub.typeclass.show(sub.value))
       }
+
+  def filterUnwantedAnns(anns: IArray[Any]): IArray[Any] =
+    anns
+      .filterNot(_.isInstanceOf[scala.SerialVersionUID])
+      .filterNot(_.isInstanceOf[scala.annotation.showAsInfix])
```
Comment (adamw) [inline src/examples/show.scala]: shouldn't we filter out these annotations from `inheritedAnns` in the first place?

File: src/core/magnolia.scala [hunk 6]
```diff
@@ -69,6 +72,39 @@ trait CommonDerivation[TypeClass[_]]:
...
+        ) ::
+          getParams_[T, ltail, ptail](
+            annotations,
+            inheritedAnnotations,
+            typeAnnotations,
+            repeated,
+            idx + 1
+          )
+
+  // for backward compatibility with v1.0.0
```
Comment (adamw) [inline src/core/magnolia.scala:107]: can this impl call getParams_ to avoid code duplication?

File: src/core/interface.scala [hunk 7]
```diff
@@ -74,6 +118,31 @@ abstract class CaseClass[Typeclass[_], Type](
...
+      typeAnnotations
+    ):
+      type PType = P
+      def default: Option[PType] = defaultVal.value
+      def typeclass = cbn.value
+      override def inheritedAnnotations = inheritedAnns
+      def deref(value: Type): P =
+        value.asInstanceOf[Product].productElement(idx).asInstanceOf[P]
+
+  // for backward compatibility with v1.0.0
```
Comment (adamw) [inline src/core/interface.scala:145]: same here, I'd just call param with the default inhertiedAnns value

File: src/core/macro.scala [hunk 8]
```diff
@@ -193,3 +144,100 @@ object Macro:
...
+    def anns: Expr[List[Any]] =
+      Expr.ofList {
+        tpe.typeSymbol.annotations
+          .filter(filterAnnotation)
+          .map(_.asExpr.asInstanceOf[Expr[Any]])
+      }
+
+    def inheritedAnns: Expr[List[Any]] =
+      Expr.ofList {
+        tpe.baseClasses
```
Comment (adamw) [inline src/core/macro.scala:162]: maybe we should filter the base classes here as well?>

File: src/examples/show.scala (hunk 5 above)
Comment (kubinio123) [reply src/examples/show.scala]: yep, after filtering out `Object` and `scala*` we don't need to filter these annotations

File: src/core/magnolia.scala (hunk 6 above)
Comment (kubinio123) [reply src/core/magnolia.scala:107]: yes, missed that some how

**Comment by adamw** [comment]
Thanks @kubinio123 @KacperFKorban  :)

**Pull request merged**